src/OrbitGl/QtTextRenderer.cpp

```diff
@@ -96,18 +100,29 @@ void QtTextRenderer::RenderLayer(QPainter* painter, float layer) {
     painter->setFont(font);
     painter->setPen(QColor(text_entry.formatting.color[0], text_entry.formatting.color[1],
                            text_entry.formatting.color[2], text_entry.formatting.color[3]));
+
+    auto stencil = manager.GetGroupState(group.name).stencil;
+    if (stencil.enabled) {
+      Vec2i stencil_screen_pos = viewport_->WorldToScreen(Vec2(stencil.pos[0], stencil.pos[1]));
+      Vec2i stencil_screen_size = viewport_->WorldToScreen(Vec2(stencil.size[0], stencil.size[1]));
+      painter->setClipRect(QRect(stencil_screen_pos[0], stencil_screen_pos[1],
+                                 stencil_screen_size[0], stencil_screen_size[1]));
+      painter->setClipping(true);
+    } else {
+      painter->setClipping(false);
+    }
+
     painter->drawText(text_entry.x, text_entry.y, text_entry.w, text_entry.h, Qt::AlignCenter,
                       text_entry.text);
   }
 }
 
-std::vector<float> QtTextRenderer::GetLayers() const {
-  std::vector<float> result(stored_text_.size());
-  int index = 0;
-  for (const auto& [layer_z, unused_text] : stored_text_) {
-    result[index++] = layer_z;
+std::vector<BatchRenderGroupId> QtTextRenderer::GetRenderGroups() const {
+  std::vector<BatchRenderGroupId> result;
+  result.reserve(stored_text_.size());
+  for (const auto& [group, unused_text] : stored_text_) {
+    result.push_back(group);
   }
-  std::sort(result.begin(), result.end());
```

Comment:
Just making sure: Not sorting them is fine because they will be sorted before rendering anyway, right?

Reply:
Yes. I've removed this for consistency with the `Batcher`, which returns the groups unsorted.